wgpu-hal/src/metal/adapter.rs

```diff
@@ -555,13 +562,27 @@ impl super::PrivateCapabilities {
             msg_send![process_info, operatingSystemVersion]
         };
 
-        let os_is_mac = device.supports_feature_set(MTLFeatureSet::macOS_GPUFamily1_v1);
-        // Metal was first introduced in OS X 10.11 and iOS 8. The current version number of visionOS is 1.0.0. Additionally,
-        // on the Simulator, Apple only provides the Apple2 GPU capability, and the Apple2+ GPU capability covers the capabilities of Apple2.
-        // Therefore, the following conditions can be used to determine if it is visionOS.
-        // https://developer.apple.com/documentation/metal/developing_metal_apps_that_run_in_simulator
-        let os_is_xr = version.major < 8 && device.supports_family(MTLGPUFamily::Apple2);
-        let family_check = os_is_xr || version.at_least((10, 15), (13, 0), os_is_mac);
+        let os_type = {
+            // Metal was first introduced in OS X 10.11 and iOS 8. The current version number of visionOS is 1.0.0. Additionally,
+            // on the Simulator, Apple only provides the Apple2 GPU capability, and the Apple2+ GPU capability covers the capabilities of Apple2.
+            // Therefore, the following conditions can be used to determine if it is visionOS.
+            // https://developer.apple.com/documentation/metal/developing_metal_apps_that_run_in_simulator
+            let os_is_xr = version.major < 8 && device.supports_family(MTLGPUFamily::Apple2);
+            let os_is_mac = device.supports_feature_set(MTLFeatureSet::macOS_GPUFamily1_v1);
+            let os_is_tvos = device.supports_feature_set(MTLFeatureSet::tvOS_GPUFamily1_v1);
+            if os_is_xr {
+                super::OsType::VisionOs
+            } else if os_is_mac {
+                super::OsType::Macos
+            } else if os_is_tvos {
+                super::OsType::Tvos
+            } else {
+                super::OsType::Ios
+            }
+        };
```

Comment:
I think we should pull this all into a OsType::new or such. Seems to be well encapsulateable.